logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -212,23 +266,104 @@ private void innerWriteEntry(DLQEntry entry) throws IOException {
+        }
+    }
+
+    private void updateOldestSegmentReference() throws IOException {
+        oldestSegmentPath = listSegmentPaths(this.queuePath).sorted().findFirst();
+        if (!oldestSegmentPath.isPresent()) {
+            oldestSegmentTimestamp = Optional.empty();
+            return;
+        }
+        // extract the newest timestamp from the oldest segment
+        Optional<Timestamp> foundTimestamp = readTimestampOfLastEventInSegment(oldestSegmentPath.get());
+        if (!foundTimestamp.isPresent()) {
+            // clean also the last segment, because doesn't contain a timestamp (corrupted maybe)
+            // or is not present anymore
+            oldestSegmentPath = Optional.empty();
+        }
+        oldestSegmentTimestamp = foundTimestamp;
+    }
+
+    /**
+     * Extract the timestamp from the last DLQEntry it finds in the given segment.
+     * Start from the end of the latest block, and going backward try to read the next event from its start.
+     * */
+    private static Optional<Timestamp> readTimestampOfLastEventInSegment(Path segmentPath) throws IOException {
+        final int lastBlockId = (int) Math.ceil(((Files.size(segmentPath) - VERSION_SIZE) / (double) BLOCK_SIZE)) - 1;
+        byte[] eventBytes;
+        try (RecordIOReader recordReader = new RecordIOReader(segmentPath)) {
+            int blockId = lastBlockId;
+            do {
+                recordReader.seekToBlock(blockId);
+                eventBytes = recordReader.readEvent();
+                blockId--;
+            } while (eventBytes == null && blockId >= 0); // no event present in last block, try with the one before
+        } catch (NoSuchFileException nsfex) {
+            // the segment file may have been removed by the clean consumed feature on the reader side
+            return Optional.empty();
+        }
+        if (eventBytes == null) {
+            logger.warn("Cannot find a complete event into the segment file [{}], this is a DLQ segment corruption", segmentPath);
+            return Optional.empty();
+        }
+        return Optional.of(DLQEntry.deserialize(eventBytes).getEntryTime());
+    }
+
     // package-private for testing
     void dropTailSegment() throws IOException {
         // remove oldest segment
         final Optional<Path> oldestSegment = listSegmentPaths(queuePath)
                 .min(Comparator.comparingInt(DeadLetterQueueUtils::extractSegmentId));
-        if (!oldestSegment.isPresent()) {
-            throw new IllegalStateException("Listing of DLQ segments resulted in empty set during storage policy size(" + maxQueueSize + ") check");
+        if (oldestSegment.isPresent()) {
+            final Path beheadedSegment = oldestSegment.get();
+            deleteTailSegment(beheadedSegment, "exceeded retained size");
+        } else {
+            logger.info("Listing of DLQ segments resulted in empty set during storage policy size({}) check", maxQueueSize);
```

Comment:
Does this only get called when the queue size is exceeded? I'm wondering what a more descriptive error message would look like here - maybe something along the lines of

```suggestion
            logger.info("Queue size {} exceeded, but no complete DLQ segments found", maxQueueSize);
```

Reply:
More user friendly and less developer oriented

Reply:
Fixed by a349277b0582114b1dcd6eebf5a923e68620fba6